File: C++/arrays/max_subarray_sum_circular.cpp
/**
 * @file max_subarray_sum_circular.cpp
 * @brief Finds the maximum subarray sum in a circular array.
 *
 * Problem:
 * Given a circular integer array `nums`, find the maximum possible sum of a non-empty subarray.
 * A circular array means that the end of the array connects back to the start.
 *
 * Example:
 * Input: nums = [5, -3, 5]
 * Output: 10
 * Explanation: The subarray [5, 5] forms a circular subarray with sum = 10.
 *
 * Approach:
 * 1. Use Kadane’s algorithm to find:
 *    - The maximum subarray sum (non-circular).
 *    - The minimum subarray sum (to compute the circular case).
 * 2. For the circular case, total_sum - min_subarray_sum gives the max circular sum.
 * 3. Return the maximum of the two cases.
 *
 * Edge Case:
 * - If all numbers are negative, the circular sum becomes 0 (invalid), 
 *   so return the maximum element (non-circular result).
 *
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int maxSubarraySumCircular(vector<int>& nums) {
        const int INF = 1 << 30;
        int minPrefixSum = 0, maxPrefixSum = -INF;
        int maxSubarraySum = -INF, minSubarraySum = INF;
        int currentPrefixSum = 0;

        for (int num : nums) {
            currentPrefixSum += num;

            // Non-wrapping case: max(prefix[j] - prefix[i])
            maxSubarraySum = max(maxSubarraySum, currentPrefixSum - minPrefixSum);

            // Wrapping helper: min(prefix[j] - prefix[i])
            minSubarraySum = min(minSubarraySum, currentPrefixSum - maxPrefixSum);

            // Update prefix boundaries
            minPrefixSum = min(minPrefixSum, currentPrefixSum);
            maxPrefixSum = max(maxPrefixSum, currentPrefixSum);
        }

        // Total sum of array
        int totalSum = currentPrefixSum;

        // If all numbers are negative, return non-circular result
        if (totalSum == minSubarraySum) return maxSubarraySum;

        // Max of non-wrapping and wrapping case
        return max(maxSubarraySum, totalSum - minSubarraySum);
    }
};

/**
 * Example usage:
 */
int main() {
    Solution sol;
    vector<int> nums = {5, -3, 5};
    cout << "Maximum Circular Subarray Sum: " 
         << sol.maxSubarraySumCircular(nums) << endl;
    return 0;
}

File: JavaScript/arrays/max_subarray_sum_circular.js
/**
 * @file max_subarray_sum_circular.js
 * @brief Finds the maximum subarray sum in a circular array.
 *
 * Problem:
 * Given a circular integer array `nums`, find the maximum possible sum of a non-empty subarray.
 * A circular array means that the end of the array connects back to the start.
 *
 * Example:
 * Input: nums = [5, -3, 5]
 * Output: 10
 * Explanation: The subarray [5, 5] forms a circular subarray with sum = 10.
 *
 * Approach (Kadane-based, explained):
 * 1. Use Kadane’s algorithm to compute the maximum subarray sum for the non-circular case:
 *    - Keep track of `maxEndingHere` (best subarray that ends at current index)
 *      and `maxSoFar` (best over all indices).
 * 2. Use a modified Kadane’s algorithm to compute the minimum subarray sum:
 *    - Keep track of `minEndingHere` (worst subarray that ends at current index)
 *      and `minSoFar` (worst over all indices).
 * 3. Let totalSum be the sum of all elements.
 *    - The best circular subarray that wraps uses all elements except the minimum subarray:
 *      `circularBest = totalSum - minSoFar`.
 * 4. The answer is the maximum of:
 *      - `maxSoFar` (non-wrapping best), and
 *      - `circularBest` (wrapping best), except when `circularBest` would be 0 because
 *        that means `minSoFar == totalSum` (i.e., all numbers are part of the min subarray),
 *        which happens when all elements are negative. In that case we should return `maxSoFar`.
 *
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */

'use strict';

/**
 * Returns the maximum subarray sum for a circular array.
 *
 * @param {number[]} nums - input array (non-empty)
 * @returns {number} maximum circular subarray sum
 */
function maxSubarraySumCircular(nums) {
  // Basic Kadane for maximum subarray sum
  let maxEndingHere = nums[0];
  let maxSoFar = nums[0];

  // Modified Kadane for minimum subarray sum
  let minEndingHere = nums[0];
  let minSoFar = nums[0];

  // Total sum of array
  let totalSum = nums[0];

  // Iterate from second element onwards
  for (let i = 1; i < nums.length; ++i) {
    const x = nums[i];
    // Standard Kadane update for max
    maxEndingHere = Math.max(x, maxEndingHere + x);
    maxSoFar = Math.max(maxSoFar, maxEndingHere);

    // Kadane-like update for min (flip the logic)
    minEndingHere = Math.min(x, minEndingHere + x);
    minSoFar = Math.min(minSoFar, minEndingHere);

    // Accumulate total sum
    totalSum += x;
  }

  // If all numbers are negative, maxSoFar is the maximum element (best non-wrapping)
  // In that case totalSum === minSoFar (because the minimum subarray is the entire array),
  // and totalSum - minSoFar would be 0 which is invalid since subarray must be non-empty.
  if (totalSum === minSoFar) {
    return maxSoFar;
  }

  // Otherwise return the maximum between non-wrapping and wrapping cases
  return Math.max(maxSoFar, totalSum - minSoFar);
}

/**
 * Example usage:
 */
const nums = [5, -3, 5];
console.log("Maximum Circular Subarray Sum:", maxSubarraySumCircular(nums)); // Expected: 10

// Export for Node.js or module systems (optional)
// module.exports = { maxSubarraySumCircular };
